staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go

```diff
@@ -318,12 +319,90 @@ func (e *Store) Create(ctx genericapirequest.Context, obj runtime.Object) (runti
 	return out, nil
 }
 
+func (e *Store) CreateInitialized(ctx genericapirequest.Context, obj runtime.Object) (runtime.Object, error) {
+	out, err := e.Create(ctx, obj)
+	if err != nil {
+		return nil, err
+	}
+
+	// return early if we don't have initializers, or if they've completed already
+	accessor, err := meta.Accessor(out)
+	if err != nil {
+		return out, nil
+	}
+	initializers := accessor.GetInitializers()
+	if initializers == nil {
+		return out, nil
+	}
+	if result := initializers.Result; result != nil {
+		return nil, kubeerr.FromObject(result)
+	}
+
+	key, err := e.KeyFunc(ctx, accessor.GetName())
+	if err != nil {
+		return nil, err
+	}
+	w, err := e.Storage.Watch(ctx, key, accessor.GetResourceVersion(), storage.SelectionPredicate{
+		Label: labels.Everything(),
+		Field: fields.Everything(),
+
+		IncludeUninitialized: true,
+	})
+	if err != nil {
+		return nil, err
+	}
+	defer w.Stop()
+
+	latest := out
+	ch := w.ResultChan()
+	for {
+		select {
+		case event, ok := <-ch:
+			if !ok {
+				// TODO: should we just expose the partially initialized object?
+				return nil, kubeerr.NewServerTimeout(e.QualifiedResource, "create", 0)
+			}
+			switch event.Type {
+			case watch.Deleted:
+				if latest = event.Object; latest != nil {
+					if accessor, err := meta.Accessor(latest); err == nil {
+						if initializers := accessor.GetInitializers(); initializers != nil && initializers.Result != nil {
+							// initialization failed, but we missed the modification event
+							return nil, kubeerr.FromObject(initializers.Result)
+						}
+					}
+				}
+				return nil, kubeerr.NewInternalError(fmt.Errorf("object deleted while waiting for creation"))
+			case watch.Error:
+				if status, ok := event.Object.(*metav1.Status); ok {
+					return nil, &kubeerr.StatusError{ErrStatus: *status}
+				}
+				return nil, kubeerr.NewInternalError(fmt.Errorf("unexpected object in watch stream, can't complete initialization %T", event.Object))
```

Comment:
On a watch disconnection we'll retry, and if we retry and the version is too old, we get a `Gone` (410).  So to get to this line something unexpected has to happen.

In general, a legacy client doesn't expect to retry a create.  We cannot safely send a retriable status code, since the object has already been created and a retry would be return already exists.  It's really not much different than a timeout